#36 Implements an option to ignore files, specified by user while scanning

ashutoshsaboo opened the pull request:
Signed-off-by: Ashutosh Saboo <ashutosh.saboo96@gmail.com>

This resolves Issue #36 . A similar PR seems to resolve this, but an additional condition needs to be added in the `resource_paths()` method while choosing the files to be ignored during the scanning process. 

This PR resolves #36 completely. I've read the discussion on the issue thread, and this commit adheres to all the specified conditions. 

@pombredanne Sir, please have a look at this. Does this seem ready to merge? Thanks!

pombredanne commented:
> Does this seem ready to merge?

Quite likely not: without even looking at the code I can see there are no unit tests.

pombredanne commented on src/scancode/cli.py:
```diff
@@ -229,14 +229,18 @@ def validate_formats(ctx, param, value):
 @click.option('--max-memory', is_flag=False, default=DEFAULT_MAX_MEMORY, type=int, show_default=True, help='Stop scanning a file if scanning requires more than a maximum amount of memory in megabytes.')
 
+#--ignore accounts for ignoring user-specified files during the scanning process.
+@click.option('--ignore', show_default=True, default='', metavar='<pattern>', multiple=True, help='Ignore files mentioned with --ignore while scanning.')
```
`multiple=True` is the way to go. But you need to find a better help text. @sschuberth any input?

sschuberth replied:
Yeah, I don't think repeating the option name as part of its description it a good idea. Maybe

> Ignore the specified file patterns when scanning.

?

ashutoshsaboo replied:
Resolved in the latest commit. :+1:

pombredanne commented on src/scancode/cli.py:
```diff
@@ -278,6 +282,8 @@ def scancode(ctx,
     scanners = OrderedDict(zip(possible_scans.keys(), zip(possible_scans.values(), scan_functions)))
 
     scans_cache_class = get_scans_cache_class()
+    #Forms a dictionary of inputs provided with --ignore
```
Is this comment useful?

ashutoshsaboo replied:
Resolved in the latest commit. :+1:

pombredanne commented on src/scancode/cli.py:
```diff
@@ -278,6 +282,8 @@ def scancode(ctx,
 
     scans_cache_class = get_scans_cache_class()
+    #Forms a dictionary of inputs provided with --ignore
+    user_ignore = dict((inputs, 'User ignore files while scanning: Supplied by --ignore') for inputs in ignore)
```
What does `'User ignore files while scanning: Supplied by --ignore'` mean here?

ashutoshsaboo replied:
Resolved in the latest commit. :+1:

pombredanne commented on src/scancode/cli.py:
```diff
@@ -289,6 +295,8 @@ def scancode(ctx,
                                     processes=processes,
                                     timeout=timeout, max_memory=max_memory,
                                     diag=diag,
+                                    #Passing a new argument to scan, to ignore files specified by user
```
Is this comment needed?

ashutoshsaboo replied:
Resolved in the latest commit. :+1:

pombredanne commented on src/scancode/cli.py:
```diff
@@ -289,6 +295,8 @@ def scancode(ctx,
                                     timeout=timeout, max_memory=max_memory,
                                     diag=diag,
+                                    #Passing a new argument to scan, to ignore files specified by user
+                                    ignore = user_ignore,
```
what about existing standard ignores? would they be ignored?

ashutoshsaboo replied:
Nope. They mustn't be ignored for stability as I mentioned below as well. :+1:

pombredanne commented on src/scancode/cli.py:
```diff
@@ -357,7 +366,8 @@ def scan(input_path,
     # maxtasksperchild helps with recycling processes in case of leaks
     pool = get_pool(processes=processes, maxtasksperchild=1000)
-    resources = resource_paths(input_path)
+    #Passing a new argument to resource_paths to count for ignoring the files
```
Is this comment needed? IMHO not.

ashutoshsaboo replied:
Resolved in the latest commit. :+1:

pombredanne commented on src/scancode/cli.py:
```diff
@@ -518,21 +528,27 @@ def _scanit(paths, scanners, scans_cache_class, diag,
     base_is_dir = filetype.is_dir(base_path)
     len_base_path = len(base_path)
-    ignored = partial(ignore.is_ignored, ignores=ignore.ignores_VCS, unignores={})
+    user_ignore.update(ignore.ignores_VCS)
```
OK, now I get it: you are merging with VCS ignores in. This should be stated in the CLI help then. Is there a way needed to bypass the VCS ignores? Not sure. I think this is fine.

ashutoshsaboo replied:
I feel we must not bypass the VCS ignores, since they contain the minimum possible details which must be ignored as well. So, i guess we must only merge the user_ignores with the default VCS ignores for stability as well.

pombredanne commented on src/scancode/cli.py:
```diff
@@ -518,21 +528,27 @@ def _scanit(paths, scanners, scans_cache_class, diag,
-    ignored = partial(ignore.is_ignored, ignores=ignore.ignores_VCS, unignores={})
+    user_ignore.update(ignore.ignores_VCS)
+    #Checks if any argument is supplied with --ignore, and if provided, then ignores the files
+    if any(user_ignore):
```
Why `if any(user_ignore)` ? `if user_ignore` is plenty enough.
Again the comments are not needed here.

ashutoshsaboo replied:
Resolved in the latest commit. :+1:

pombredanne commented on src/scancode/cli.py:
```diff
@@ -518,21 +528,27 @@ def _scanit(paths, scanners, scans_cache_class, diag,
+    #Checks if any argument is supplied with --ignore, and if provided, then ignores the files
+    if any(user_ignore):
+        ignored = partial(ignore.is_ignored, ignores=user_ignore, unignores={})
+    else:
```
Is this else really needed at all? I think this can be simplified.
Your `user_ignore` should contain only the vsc or vsc + CLI inputs. So there should not be a need for tow partials.

ashutoshsaboo replied:
Resolved in the latest commit. :+1:

ashutoshsaboo commented:
Hi @pombredanne ! Thanks for your input. I have resolved everything and added the unit tests as well. Just a thing, I don't know why my latest two tests seem to fail, while they work correctly when I execute CLI on my local system. Can you check once?

Have I passed arguments, in some wrong manner in the new unit tests that I added?

ashutoshsaboo commented:
@pombredanne Can you review this PR as well?

pombredanne commented on src/scancode/cli.py:
```diff
@@ -229,14 +229,18 @@ def validate_formats(ctx, param, value):
              license_score, license_text, only_findings, strip_root,
              format, verbose, quiet, processes,
-             diag, timeout, max_memory,
+             diag, timeout, max_memory,ignore,
```
pep8 spacing please

pombredanne commented on src/scancode/cli.py:
```diff
@@ -229,14 +229,18 @@ def validate_formats(ctx, param, value):
-             diag, timeout, max_memory,
+             diag, timeout, max_memory,ignore,
              *args, **kwargs):
+    #Passing a new argument ignore to scancode() to account for ignoring files specified by user
```
This is not needed

pombredanne commented on src/scancode/cli.py:
```diff
@@ -278,6 +282,7 @@ def scancode(ctx,
     scanners = OrderedDict(zip(possible_scans.keys(), zip(possible_scans.values(), scan_functions)))
 
     scans_cache_class = get_scans_cache_class()
+    user_ignore = dict((inputs, 'Ignore files while scanning, that are supplied by --ignore') for inputs in ignore)
```
Use a dict comprehension instead. Do not repeat the help text of the command option

pombredanne commented on src/scancode/cli.py:
```diff
@@ -313,6 +319,7 @@ def scan(input_path,
          processes=1, timeout=DEFAULT_TIMEOUT, max_memory=DEFAULT_MAX_MEMORY,
          diag=False,
          scans_cache_class=None,
+         ignore={},
```
NEVER ever use a mutable collection in an arg. This is a global

pombredanne replied:
?

pombredanne commented on src/scancode/cli.py:
```diff
@@ -518,21 +525,23 @@ def _scanit(paths, scanners, scans_cache_class, diag,
 
 
-def resource_paths(base_path):
+def resource_paths(base_path, user_ignore={}):
```
No globals please!

pombredanne commented on tests/scancode/test_cli.py:
```diff
@@ -452,3 +452,25 @@ def test_scan_can_run_from_other_directory():
         print(stderr)
     assert rc == 0
     check_scan(test_env.get_test_loc(expected_file), result_file, strip_dates=True, regen=False)
+
```
You need to add proper unit tests to your function alone and not only CLI integration tests

ashutoshsaboo replied:
@pombredanne Yes I'll add them as well. Sorry for the inconvenience. Actually, I was confused, as to why the tests weren't passing, even though it seemed to work correctly when I executed the Scancode CLI locally though. That was my confusion actually.

pombredanne commented:
> Actually, I was confused, as to why the tests weren't passing, even though it seemed to work correctly when I executed the Scancode CLI locally though.

Well this is why we run tests on Linux/Windows/macOS on the CI side to keep us honest. This is to avoid this: https://blog.codinghorror.com/the-works-on-my-machine-certification-program/ ;)

pombredanne commented:
@ashutoshsaboo is this something you intend to complete?

ashutoshsaboo commented:
@pombredanne Yes, actually caught up with exams. Hence, will try to take time, and resolve this soon. :smile:

pombredanne commented on src/scancode/cli.py:
```diff
@@ -229,14 +229,18 @@ def validate_formats(ctx, param, value):
 @click.option('--timeout', is_flag=False, default=DEFAULT_TIMEOUT, type=int, show_default=True, help='Stop scanning a file if scanning takes longer than a timeout in seconds.')
 @click.option('--max-memory', is_flag=False, default=DEFAULT_MAX_MEMORY, type=int, show_default=True, help='Stop scanning a file if scanning requires more than a maximum amount of memory in megabytes.')
 
+#--ignore accounts for ignoring user-specified files during the scanning process.
```
This comment is not needed.

ashutoshsaboo replied:
@pombredanne I updated the commit on PR #544 actually. Not this one yet. :smile:

pombredanne replied:
I am rather confused now as to which PR to review?

pombredanne replied:
?

pombredanne commented:
Note that beside the fact that the tests fail, you need to rebase too as your branch went stale

pombredanne commented:
@ashutoshsaboo I am closing this because I went with @yasharmaster #538 
But I really appreciate all the efforts you did put into this. Thanks!
